packages/react-server/src/ReactFizzServer.js

```diff
@@ -2035,29 +2061,45 @@ function flushCompletedQueues(
   request: Request,
   destination: Destination,
 ): void {
+  let allComplete = false;
   beginWriting(destination);
   try {
     // The structure of this is to go through each queue one by one and write
     // until the sink tells us to stop. When we should stop, we still finish writing
     // that item fully and then yield. At that point we remove the already completed
     // items up until the point we completed them.
 
-    // TODO: Emit preloading.
-
-    // TODO: It's kind of unfortunate to keep checking this array after we've already
-    // emitted the root.
+    let i;
     const completedRootSegment = request.completedRootSegment;
-    if (completedRootSegment !== null && request.pendingRootTasks === 0) {
-      flushSegment(request, destination, completedRootSegment);
-      request.completedRootSegment = null;
-      writeCompletedRoot(destination, request.responseState);
+    if (completedRootSegment !== null) {
+      if (request.pendingRootTasks === 0) {
+        // Flush the prelude if it exists
+        const prelude = request.prelude;
+        for (i = 0; i < prelude.length; i++) {
+          // we expect the prelude to be tiny and will ignore backpressure
+          writeChunk(destination, prelude[i]);
+        }
+        prelude.length = 0;
+
+        // Flush resources
+        writeResources(destination, request.resources);
```

Comment:
If we have a `completedRootSegment` it is safe to flush the prelude (we should consider barring use of the prelude after the root segment has flushed. This currently can't / doesn't happen but a future change could alter that)

Once we have flushed the prelude we can flush resources and then the segment.